### ISSUE TITLE
Difficulty getting deepspeech-sys to compile for examples

### ISSUE THREAD

**Issue by quietlychris** [RustAudio/deepspeech-rs#20]
Hello,

I've run into a few issues with getting started using the examples, and to check if either something's not working on with the repository, or if I'm making a mistake that other people might run into as well. 

I currently have the deepspeech-0.6.0-models (extracted from the tarball downloaded from the release page) located in a directory along the lines of `~/rust-projects/speech-recognition`. I interpreted the instructions included in the `README` such that I have added the following lines to my `.bashrc` file located in my home directory, and used `$ source ~/.bashrc` to make sure that that has been reloaded into my current shell.
```
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/rust-projects/speech-recognition/deepspeech-0.6.0-models
export LIBRARY_PATH=$LIBRARY_PATH:~/rust-projects/speech-recognition/deepspeech-0.6.0-models
 ```
I cloned this repository into the same directory as the model, added in an audio sample, and now have a directory that looks like the following: 
```
speech-recognition
├── audio_sample.wav
├── deepspeech-0.6.0-models
└── deepspeech-rs
```
However, when I run the following, I'm unable to finish compiling, always running into an error with `deepspeech-sys`. Here's the output of the shell at that stage:
```
chrism@chrism:~/rust-projects/speech-recognition/deepspeech-rs$ cargo run --release --example client ../deepspeech-0.6.0-models ../audio_sample.wav 
   Compiling deepspeech-sys v0.6.0 (/home/chrism/rust-projects/speech-recognition/deepspeech-rs/sys)
error: failed to run custom build command for `deepspeech-sys v0.6.0 (/home/chrism/rust-projects/speech-recognition/deepspeech-rs/sys)`

Caused by:
  process didn't exit successfully: `/home/chrism/rust-projects/speech-recognition/deepspeech-rs/target/release/build/deepspeech-sys-05fceb855d5921c7/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-link-lib=deepspeech
cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
cargo:warning=set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

--- stderr
error: header 'deepspeech/native_client/deepspeech.h' does not exist.
thread 'main' panicked at 'Couldn't generate bindings: ()', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```
I think this has something to do with my directory structures or naming conventions where something in my system isn't being located properly, but the things I've tried to resolve them (switching around directory names and my naming convention for the library paths) haven't worked. It's also possible I could be missing a system-level component wrt to LLVM that's expected to be present, but isn't. I'm currently running `rustc-1.40-nightly` on Ubuntu 18.04 LTS. 

Any tips you might be able to offer for getting things up and running would be very much appreciated!

**Comment by est31** [comment]
It can't seem to find the `llvm-config` executable. Not sure which component needs it, but you can get it by installing the `llvm` package: `sudo apt install llvm`. Hope this helps.

**Comment by quietlychris** [comment]
Thanks for the response! The following two lines of the error has been resolved, 
```
cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
cargo:warning=set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable
``` 
but I'm still seeing the 
```
cargo:rustc-link-lib=deepspeech
``` 
issue, as well as the entirety of the `stderr` output. Any thoughts?

**Comment by est31** [comment]
Oh you might need the submodules: `git submodule update --init`.

**Comment by futurechimp** [comment]
@quietlychris I needed to `sudo apt install llvm clang` to get past these errors. However, I think my `LD_LIBRARY_PATH` and `LIBRARY_PATH` are not correct, and I think I'm confused about where those are supposed to point (I've tried quite a few different things :) ).

My current output is: 

```
$ cargo run --release --example client ../deepspeech-0.6.0-models/ ../audio/2830-3980-0043.wav 
   Compiling deepspeech v0.6.0
error: linking with `cc` failed: exit code: 1
<snip: big stack of directories searched>
"-Wl,--end-group" "-Wl,-Bdynamic" "-ldeepspeech" "-lutil" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /usr/bin/ld: cannot find -ldeepspeech
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: could not compile `deepspeech`.
```

I think I'm a bit unclear about what the README is telling me to do, I've never compiled a Rust binding before. 

The README says:

> Add the directory where the precompiled components lie (the DeepSpeech checkout) to your LD_LIBRARY_PATH and LIBRARY_PATH environment variables.

Does "the DeepSpeech checkout" in the sentence above mean the DeepSpeech code that results from `git submodule update --init`? 

Or does it refer to this part? 

> Obtain the Deepspeech native_client library. The release announcement contains precompiled libraries for various targets.

It seems to me that the checkout in `sys` already has that (?).

**Comment by quietlychris** [comment]
@futurechimp Unfortunately, I got a bit distracted over the last few days, and didn't have a chance to figure out a fix for this problem. 

I'm currently at the same point that you are: installing `llvm` fixed all but the linker error. I've also tried placing a few different directories (including both `sys/deepspeech` and `sys/deepspeech/native_client`) into `LD_LIBRARY_PATH` and `LIBRARY_PATH`, and as far as I can tell using `$ echo $LD_LIBRARY_PATH`, they have been added successfully. I haven't found that using the `git submodule update --init` command changed the error regarding the linker. 

It does seem like the `sys` should have those parts, and as far as I can tell, it doesn't have anything to do with system-level access to some other DeepSpeech repository that's not installed (using the Mozilla DeepSpeech repository, and the `deepspeech-0.6.0-models` directory downloaded from there, I'm able to use the command-line client `$ deepspeech --model deepspeech-0.6.0-models/output_graph.pbmm --audio audio_sample.wav` to access the model and successfully run an inference). 

I'm not terribly familiar with the specification of a `build.rs` script, but it seems like the issue might be somehow located there? From what I can tell, some sort of `bindings.rs` file should be created. However, it seems like the `$OUT_DIR` variable called in that script is pulled from the build environment, and using `echo $OUT_DIR` doesn't bring anything up for me, so that might be an issue. In any case, after using `cargo build` in the `sys` directory, the only file with any sort of similar name is found at `deepspeech/native_client/javascript/binding.gyp`, which doesn't seem particularly relevant. I'm also unable to really find anything that looks like the kind of `bindgen` output I'd expect to see for an autogenerated C++ to Rust FFI header. I could certainly be mistaken about that, though. 

It's good to get a confirmation that someone else has been seeing something similar, and that it's not isolated to some particular quirk of my development environment. @est31, do you have any ideas about how we might be able to resolve the aforementioned linker issue?

**Comment by est31** [comment]
> note: /usr/bin/ld: cannot find -ldeepspeech

@futurechimp this linker error is due to wrong configuration of `LD_LIBRARY_PATH` and `LIBRARY_PATH`.

> >    Add the directory where the precompiled components lie (the DeepSpeech checkout) to your LD_LIBRARY_PATH and LIBRARY_PATH environment variables.
>
> Does "the DeepSpeech checkout" in the sentence above mean the DeepSpeech code that results from git submodule update --init?

Actually no. You need to point it to the directory with the precompiled components. You should have obtained it in step 1 in the readme:

> Obtain the Deepspeech native_client library. The release announcement contains precompiled libraries for various targets.

Extract it somewhere and point `LD_LIBRARY_PATH` and `LIBRARY_PATH` to it. It's kinda my fault for not pointing out your mistake as in your initial comment you pointed them to the model dir which is wrong obviously.

The "DeepSpeech checkout" in that sentence refers to the situation if you build DeepSpeech yourself. Then the compiled library is put somewhere into your checkout. I guess that's not phrased really well.

**Comment by quietlychris** [comment]
Okay, that makes more sense--thank you! Just to re-iterate, based on the steps that allowed me to get past the linker error and compile:
1. Install the `llvm` and `clang` libraries
2. In `deepspeech-rs/sys/deepspeech/util` there is a python script named `taskcluster.py`, which should download (if it's not already present) a `native_client.tar.xz` file. You are able to specify where this file will be downloaded to.
3. Extract the contents of that tarball
4. Add the new `native_client` directory to the path of the `LD_LIBRARY_PATH` and `LIBRARY_PATH` variables
5. If not building the DeepSpeech model yourself, download the pre-built models from the Mozilla DeepSpeech release announcement (i.e. `deepspeech-0.6.0-models`).I've found the root of my project's working directory to be the most convenient place for me.
6. In the root of `deepspeech-rs`, it should now be possible to compile the project with a command along the lines of `$ cargo run --release --example client deepspeech-0.6.0-models --audio audio_sample.wav`

I'm able to compile and run the `client` examples, but am seeing a segfault occur before the inference is finished running. It's something I plan on looking into, and will update with results when I have them.

**Comment by est31** [comment]
The native client is also included in the release announcements, but using taskcluster is okay as well.

**Comment by futurechimp** [comment]
As it turns out, my problem is that I'm running Old Man Linux: Ubuntu 16.04. The `clang` and `llvm` included in Ubuntu 16.04 don't work. Trying it out on a friend's newer machine (Ubuntu 19.10) works well. The `LD_LIBRARY_PATH` stuff was a red herring, I thought I didn't have it set correctly but it was simply that the linker was going to fail no matter what I did. 

Thanks @est31 and @quietlychris!

**Comment by est31** [comment]
Alright it seems to work for both of you now. Given that #22 is merged which should make using deepspeech-rs much easier, I'll close this. If you have any further questions please file a new issue. Thanks!
